Plugins~/Src/MeshSyncClientBlender/msblenMaterialsExportHelper.cpp

```diff
@@ -0,0 +1,484 @@
+#include "msblenMaterialsExportHelper.h"
+
+#include <BKE_node.h>
+#include <BLI_utildefines.h>
+
+#include "msblenUtils.h"
+#include "MeshSync/SceneGraph/msTexture.h"
+#include "MeshSync/Utility/msMaterialExt.h"
+
+namespace blender {
+// Blender uses hardcoded string identifiers to figure out what the sockets do:
+const auto baseColorIdentifier = "Base Color";
+const auto colorIdentifier = "Color";
+const auto roughnessIdentifier = "Roughness";
+const auto metallicIdentifier = "Metallic";
+const auto normalIdentifier = "Normal";
+const auto normalStrengthIdentifier = "Strength";
+const auto surfaceIdentifier = "Surface";
+const auto emissionIdentifier = "Emission";
+const auto emissionStrengthIdentifier = "Emission Strength";
+
+const auto shaderIdentifier = "Shader";
+
+const auto displacementIdentifier = "Displacement";
+const auto heightIdentifier = "Height";
+const auto scaleIdentifier = "Scale";
+
+// Removes reroute nodes and returns the actual input node upstream.
+bNode* removeReroutes(bNode* node, const Material* mat) {
+	if (!node || node->type != NODE_REROUTE) {
+		return node;
+	}
+
+	auto tree = mat->nodetree;
+
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode == node) {
+			return removeReroutes(link->fromnode, mat);
+		}
+	}
+
+	return nullptr;
+}
+
+bNode* getNodeConnectedToSocket(bNodeSocket* socket) {
+	if (socket->link) {
+		return socket->link->fromnode;
+	}
+
+	return nullptr;
+}
+
+// For mix shaders, we pass through the first input that has a connection.
+bNode* handleMixShader(const Material* mat, bNode* bsdf) {
+	if (!bsdf || bsdf->type != SH_NODE_MIX_SHADER)
+		return bsdf;
+
+	for (auto inputSocket : list_range((bNodeSocket*)bsdf->inputs.first)) {
+		if (STREQ(inputSocket->name, shaderIdentifier)) {
+			bNode* connectedBSDF = removeReroutes(getNodeConnectedToSocket(inputSocket), mat);
+			if (connectedBSDF)
+			{
+				return connectedBSDF;
+			}
+		}
+	}
+
+	return bsdf;
+}
+
+bool getBSDFAndOutput(const Material* mat, bNode*& bsdf, bNode*& output) {
+	// Find BSDF connected to an output node:
+	auto tree = mat->nodetree;
+	for (auto link : list_range((bNodeLink*)tree->links.first)) {
+		if (link->tonode &&
+			link->tonode->type == SH_NODE_OUTPUT_MATERIAL &&
+			STREQ(link->tosock->identifier, surfaceIdentifier)) {
+			bsdf = removeReroutes(link->fromnode, mat);
+
+			bsdf = handleMixShader(mat, bsdf);
+
+			output = link->tonode;
+			return bsdf && output;
+		}
+	}
+
+	return false;
+}
+
+bNodeSocket* getInputSocket(bNode* node, const char* socketName) {
+	for (auto inputSocket : list_range((bNodeSocket*)node->inputs.first)) {
+		if (STREQ(inputSocket->name, socketName)) {
+			return inputSocket;
+		}
+	}
+
+	return nullptr;
+}
+
+int msblenMaterialsExportHelper::exportTexture(const std::string& path, ms::TextureType type) const
+{
+	return m_texture_manager->addFile(path, type);
+}
+
+void msblenMaterialsExportHelper::handleImageNodeWithAssignedImage(ms::TextureType& textureType,
```

Comment:
This function does 2 things: 
1. invokes the texture handler on the texture, which might be either in the blendfile or somewhere in the filesystem
2. invokes the color handler

Can these 2 responsibilities be broken down into 2 smaller functions? It should allow us to reduce some nesting by having some guard checks at the top with early exits

Reply:
Or probably even better, extract the code that exports an image from a blend file to it's own function

Reply:
Done